### PR TITLE
Displayed plasma.graph with matplotlib

### PR DESCRIPTION
**Pull request by Chinmay4400**
Updated quickstart to show plasma graph from a png file.

### CODE REVIEW

**Comment by wkerzendorf** [comment]
so we don't want to store the png with it but we want to generate it automatically.

File: docs/quickstart/quickstart.ipynb [hunk 1]
```diff
@@ -26,14 +26,19 @@
      "name": "stderr",
```
Comment (wkerzendorf) [inline docs/quickstart/quickstart.ipynb]: <p>and this is already in the notebook - so we do not need the png file - right. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tardis-sn/tardis/pull/1042/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (Chinmay4400) [reply docs/quickstart/quickstart.ipynb]: The png file doesn't have to be stored in the github repo.

Comment (Chinmay4400) [reply docs/quickstart/quickstart.ipynb]: should i remove it ?

Comment (rohitp934) [reply docs/quickstart/quickstart.ipynb]: The png file will be created each time a user runs the notebook though, but as long as the png file is not on the repo it should be fine right @wkerzendorf

Comment (wkerzendorf) [reply docs/quickstart/quickstart.ipynb]: @rohitp934 - this is something that @Chinmay4400 is working on - please make your own implementation if you are interested in this project.

Comment (wkerzendorf) [reply docs/quickstart/quickstart.ipynb]: @Chinmay4400 well think about what we are trying to accomplish - it is already showing it - so do we need the png?

Comment (Chinmay4400) [reply docs/quickstart/quickstart.ipynb]: If a user runs the notebook, then it's fine. But if we view it in github, we might need the png file. Let me get back to you on this.

Comment (Chinmay4400) [reply docs/quickstart/quickstart.ipynb]: deleting the png file breaks the image in notebook in github repo

Comment (Chinmay4400) [reply docs/quickstart/quickstart.ipynb]: @wkerzendorf I deleted the image but now I can't see the image in the quickstart notebook...not without running. Makes sense as I had displayed it in markdown.

Comment (Chinmay4400) [reply docs/quickstart/quickstart.ipynb]: @wkerzendorf I have updated the quickstart notebook as required. Please take a look

Comment (wkerzendorf) [reply docs/quickstart/quickstart.ipynb]: <p>now make the image zoomable. </p>

Comment (Chinmay4400) [reply docs/quickstart/quickstart.ipynb]: double clicking the image allows zooming. Is that not what you have in mind?

**Comment by wkerzendorf** [comment]
try to make it zoomable like google maps

**Comment by Chinmay4400** [comment]
In the latest commit quickstart is updated to make a graph directly from plasma.graph which is a networkx.digraph object. 
Graph has been made interactive using %matplotlib notebook, however to properly view the graph some adjustments are required. I am working on these.
If we move forward, plotting this way, we won't require to create dot file for plotting and viz purposes.
Plasma.write_to_dot has been updated to be compatible with networkx API.

**Comment by wkerzendorf** [comment]
okay - now we also want the formula in there

**Comment by Chinmay4400** [comment]
Can you please elaborate... I'm not sure if I know about the formula you're talking about.

**Comment by Chinmay4400** [comment]
Also, I wanted to update the plasma.module to be compatible with networkx and other libraries that it imports and add a display_graph method to achieve this graph...should i make a different PR for that?

**Comment by wkerzendorf** [comment]
have a look at the pngs you were making beforehand

**Comment by wkerzendorf** [comment]
Some of the nodes have a keyword `latex_formula` that should be printed below the node. The node organisation is better in graphviz - find a way to emulate this in matplotlib.

**Comment by Chinmay4400** [comment]
@wkerzendorf I have got the formulae as edge labels, but I'm facing some problems displaying them with graph. I'll let you know further shortly.

**Comment by wkerzendorf** [comment]
okay - going in the right direction 👍 . So now use the function that puts out a tex file and compile that tex file to PDF and see what it looks like. Good luck!

**Comment by Chinmay4400** [comment]
On it

**Comment by Chinmay4400** [comment]
@wkerzendorf The dot file generated by write_to_dot is causing errors in while writing to texcode. The function works when i try it on other sample dot strings. Do you have any idea what might be going wrong?

**Comment by wkerzendorf** [comment]
@Chinmay4400 I think that sounds like a good project - figure out whats wrong the `write_to_dot` and fix that. That should be a new PR.

**Comment by Chinmay4400** [comment]
@Chinmay4400 Actually I have might have an idea
The write_to_dot function appends '$'s before and after labels. This caused problems while making the graphs too.  I'll look into it.
Is the latest commit in this branch ok?

**Comment by marxwillia** [comment]
Instead of displaying the raw latex strings as edge labels, can you get matplotlib to compile the latex and display the actual formulas?

**Comment by Chinmay4400** [comment]
@marxwillia I have started working on this. Will let you know soon.

**Comment by Chinmay4400** [comment]
Generating latex code in matplotlib can be done in the following ways :
1. Using the inbuilt [mathtext](https://matplotlib.org/3.1.3/tutorials/text/mathtext.html) in matplotlib (would require making some changes in base plasma module)
2. Having a complete installation of TeX.
Reason for this :
The plasma module uses a method plasma._build_graph :
`label = plasma.outputs_dict[input].latex_name[position]`
to generate tex code.
This code is not recognised by when we use matplotlib's mathtext

**Comment by Chinmay4400** [comment]
I can either change the latex names for plasma properties
or work with a complete installation of tex

**Comment by Chinmay4400** [comment]
```class Lines(BaseAtomicDataProperty):
    """
    Attributes
    ----------
    lines : Pandas DataFrame (wavelength, atomic_number, ion_number, f_ul, f_lu, level_number_lower,
                              level_number_upper, nu, B_lu, B_ul, A_ul, wavelength)
            All atomic lines data. Index = line_id.
    nu : Pandas DataFrame (index=line_id), dtype float
            Line frequency data
    f_lu : Pandas DataFrame (index=line_id), dtype float
            Transition probability data
    wavelength_cm: Pandas DataFrame (index=line_id), dtype float
            Line wavelengths in cm
    """
# Would like for lines to just be the line_id values
    outputs = ('lines', 'nu', 'f_lu', 'wavelength_cm') ```
    I am updating latex values to be compatible with matplotlib - This block of function is in plasma properties but no latex has been defined...should i leave it as it is... the values in output appear in graph

**Comment by wkerzendorf** [comment]
now I want you to write out the tex file and make a PDF and share it here.

**Comment by marxwillia** [comment]
Making the labels consistent with matplotlib is a good choice since we don't want to have to manage a full tex installation for tardis users.  If there are tex labels missing, try adding them where appropriate.

**Comment by Chinmay4400** [comment]
@wkerzendorf 
[test.pdf](https://github.com/tardis-sn/tardis/files/4280227/test.pdf)

**Comment by wkerzendorf** [comment]
so there are still a few bugs in there like the "textrmlevels". you can see how the graph flows from inputs to outputs which is nice. We want the same one in the matplotlib. and in your tests use the pytest temporary directory. Next is to make a similar graph in matplotlib with the same layout

**Comment by Chinmay4400** [comment]
@wkerzendorf  I used the pytest temp directory in my other PR. I have made some common changes in both PRs. Please take a look [here](https://github.com/tardis-sn/tardis/pull/1057).
I'm aware of the few bugs, those are because I was trying out different strings to see which gave the best graphs, I'll rectify them.

**Comment by Chinmay4400** [comment]
Also the 'textrm' in all labels will have to be dropped if we want the same graph in matplotlib as I pointed out earlier, so that won't be an issue anymore.

**Comment by Chinmay4400** [comment]
i also want to clean up the code a little as my local branch has gotten kind of messy. Can you merge my other PR? so I won't be making the same changes in separate branches. Thanks.

**Comment by Chinmay4400** [comment]
@marxwillia @wkerzendorf I have updated latex code and assigned latex formulae where required. The graph is updated with a slightly different colour scheme. I checked the tex graph generated after the changes made in this code and it is still the same. Please review and let me know. Thanks

**Comment by Chinmay4400** [comment]
I have removed the graph generation code from quickstart and wrapped it in a method plasma.display_plasma_graph. This method takes most parameters from users to display graph in the format they prefer. The method is complete with docstring. I have yet to write test for this method, which I think should include wether the graph generated (Networkx DiGraph) has all the nodes and labels as intended. I don't have much idea as to how can I test the graph itself considering it will be different for different set parameters, hence there isn't anything to compare the graph against. I am working on this and will soon add the test.

**Comment by wkerzendorf** [comment]
please make the graph layout to be the same as in the PDF.

**Comment by Chinmay4400** [comment]
@wkerzendorf I have read around and tried to implement some techniques but I couldn't find much on testing a function that plots a graph. I'm not sure how to tackle this issue.

**Comment by Chinmay4400** [comment]
@wkerzendorf I tried to do that but there are several problems in getting the same layout:
1. Matplotlib doesn't allow arrows to have random shape
2. Increasing the node size to completely contain the labels will require to make the graphs very large.
3. the node positioning is same as the pdf file but the edge labels get plotted automatically assuming straight line to between nodes. If we introduce a little arc in the edges, the labels still stay at their place.

**Comment by Chinmay4400** [comment]
It'd be great if you could point me in the right direction.

**Comment by marxwillia** [comment]
@chinmay0301 some of the code that you have written in this PR is not compliant with the numpydocs style guides.  Can you please fix that?

**Comment by Chinmay4400** [comment]
@marxwillia Yes Sir. Also can you please guide me further about the graph layout. I went through several plotting and graph resources but I couldn't find much on plotting LaTeX style graph in mpl. Thanks.

**Comment by wkerzendorf** [comment]
@Chinmay4400 where have you looked for the layout?

**Comment by Chinmay4400** [comment]
@wkerzendorf i have looked in networkx and pygraphviz documentations. And explored some small libraries that I mentioned in the tardis-gsoc channel. The only way i found to get this plot is as a png image. I could find some way to display that image propely, like an mpl plot if that's alright. If you could point me towards something for latex like graph plotting, it would be great.

**Comment by wkerzendorf** [comment]
There are a few layouts here: https://networkx.github.io/documentation/stable/reference/drawing.html#module-networkx.drawing.layouts - here.

**Comment by Chinmay4400** [comment]
Yes sir. I have used all methods given there. @wkerzendorf I have also posted images from those methods to tardis channel. There were only three plots that had minimal overlapping. I have posted those. Should I post those images here?

**Comment by wkerzendorf** [comment]
you could - that will make it easier for us to compare - I'm pretty sure that there is a way to do this.

**Comment by Chinmay4400** [comment]
Also these methods only change the layout of our graph. Layout is not a problem because the current layout(positioning) is same as the pdf file. But the node and edge shape and labels are different.

**Comment by Chinmay4400** [comment]
spring
spring layout

**Comment by Chinmay4400** [comment]
circular
circular layout

**Comment by Chinmay4400** [comment]
kamada_kawai
kamada-kawai

**Comment by Chinmay4400** [comment]
and the dot layout is currently displayed in the notebook.

**Comment by Chinmay4400** [comment]
@marxwillia The only possible way I see, is to make a patch in mpl that allows us to have elliptical nodes and random shaped edges. I'll have to read deeper in matplotlib docs in order to get that patch. It would require some time.

**Comment by Chinmay4400** [comment]
@wkerzendorf I'd like to explore some other visualization libraries apart from matplotlib, I'm not sure if I can achieve the pdf style plot through them, but I have some ideas to improve interactivity.
some of these libs are mentioned here 
https://mode.com/blog/python-interactive-plot-libraries

**Pull request closed without merging**